README.md
# Split Evaluator service

This services exposes a set of APIs to produce server side evaluation of flags by wrapping a NodeJS SDK.

## Usage with NodeJS

1. `nvm use`
2. `npm install`
3. `SPLITIO_API_KEY=xxxxxxx SPLITIO_EXT_API_KEY=yyyyyy SPLITIO_SERVER_PORT=7548 SPLITIO_DEBUG='on' npm start`

   SPLITIO_API_KEY could be overriden quickly, but we recommend the usage of [node-config](https://github.com/lorenwest/node-config#quick-start),
   switch the environment variable NODE_ENV. Please read the details [here](https://github.com/lorenwest/node-config#quick-start).  
   SPLITIO_EXT_API_KEY could be any value you like, against which we will compare the received Authorization header.
   SPLITIO_SERVER_PORT is the port number on which the server will run, default is 7548.  
   For logging details see the [NodeJS SDK docs](https://docs.split.io/docs/nodejs-sdk-overview#section-logging).  

## Services

    GET
      /get-treatment

    QUERY PARAMS
      key:
        This is the key used in the getTreatment call.
      bucketing-key:
        (Optional) This is the bucketing key used in the getTreatment call.
      split-name:
        This should be the name of the split you want to include in the getTreatment call.
      attributes:
        (Optional) This should be a json string of the attributes you want to include in the getTreatment call.

    EXAMPLE
      curl 'http://localhost:4444/get-treatment?key=my-customer-key&split-name=my-experiment&attributes=\{"attribute1":"one","attribute2":2,"attribute3":true\}' -H 'Authorization: {SPLITIO_EXT_API_KEY}'

    RESPONSE
      {
        "splitName": "my-experiment",
        "treatment": "on"
      }

    GET
      /get-treatments

    QUERY PARAMS
      keys:
        This is the array of keys to be used in the getTreatments call. Each key should specify a `matchingKey` 
        and a `trafficType`. You can also specify a `bucketingKey`.
      attributes:
        (Optional) This should be a json string of the attributes you want to include in the getTreatments call.

    EXAMPLE
      curl 'http://localhost:4444/get-treatments?keys=\[\{"matchingKey":"my-first-key","trafficType":"account"\},\{"matchingKey":"my-second-key","bucketingKey":"my-bucketing-key","trafficType":"user"\}\]&attributes=\{"attribute1":"one","attribute2":2,"attribute3":true\}' -H 'Authorization: {SPLITIO_EXT_API_KEY}'

    RESPONSE
      [
        { "splitName": "my-experiment", "treatment":"on" },
        { "splitName": "my-second-experiment", "treatment":"off" },
        ...
        { "splitName": "my-last-experiment", "treatment":"visa" }
      ]

    GET 
      /admin/version

    EXAMPLE 
      curl 'http//localhost:4444/admin/version' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
      
    RESPONSE
      {
        version: '1.0.2',
        sdk: 'nodejs',
        sdkVersion: '9.3.4'
      }

    GET 
      /admin/machine

    EXAMPLE 
      curl 'http//localhost:4444/admin/machine' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
      
    RESPONSE
      {
        ip: '10.0.0.125',
        name: 'machine_name'
      }
    
    GET 
      /admin/ping

    EXAMPLE 
      curl 'http//localhost:4444/admin/ping' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
      
    RESPONSE
      200: 'pong'
    
    GET 
      /admin/healthcheck

    EXAMPLE 
      curl 'http//localhost:4444/admin/healthcheck' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
      
    RESPONSE
      200: 'Split Evaluator working as expected.'
      500: 'Split evaluator engine is not evaluating traffic properly.'

    GET 
      /admin/uptime

    EXAMPLE 
      curl 'http//localhost:4444/admin/uptime' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
      
    RESPONSE
      200: '0d 23h 5m 10s'


## Running the service in Docker container

### Command to build & run the docker container :
By default te server uses the port 7548, which is the one exposed by the container.
You can use a different one by setting the `SPLITIO_SERVER_PORT` environment variable,
but their exposure of that port to the host will depend on your settings.
_You can just leave the default port and map it to whatever port you need_

*Pull the image:* `docker pull splitsoftware/split-evaluator:<VERSION>`  

*Run the container:*  

```shell
docker run -e SPLITIO_API_KEY={YOUR_API_KEY} -e SPLITIO_DEBUG='off' -e SPLITIO_SERVER_PORT=7549 -p 4444:7549 splitsoftware/split-evaluator:<VERSION>
```

**NOTE:** *SPLITIO_DEBUG & SPLITIO_SERVER_PORT are optionals*

#### Configs:
`SPLITIO_EXT_API_KEY` : Callhome will validate every request against Authorization header. This is not a Split API key but an arbitrary value.  
`SPLITIO_API_KEY` : Api-Key for you Split Environment.  
`SPLITIO_DEBUG` : (Optional) Usable for enabling/disabling the logs of the SDK. By default they are off.  
`SPLITIO_SERVER_PORT` :  TCP Port of the server inside the container.

## Build Docker Image
This command must be executed at root folder
`docker build -t splitsoftware/split-evaluator:<VERSION> . && docker build -t splitsoftware/split-evaluator:latest .`

## Push Docker Image
Before pushing image you must be logged in docker cloud. So run this command:
`docker login`
And push the image:
`docker push splitsoftware/split-evaluator:<VERSION> && docker push splitsoftware/split-evaluator:latest`

Pushing `latest` image. If tag is not explicit on `docker pull` command, the tag will be set by Docker engine as `latest`. So it is important create an image with this tag and push it.

```shell
docker build -t splitsoftware/split-evaluator:latest .
docker push splitsoftware/split-evaluator:latest
```


NEWS.txt
## 1.4.5 (October 8, 2018)

* Updating SDK version to v10.4.0 which is the latest stable to the date. Adding tilde.
* Updating Lodash and Express for vulnerability fixes.

## 1.4.4 (June 4, 2018)

* Updating SDK version to v10.3.1 which is the latest stable to the date.

## 1.4.3 (March 8, 2018)

* Updating SDK version to v10.1.0 which is the latest stable to the date.

## 1.4.2 (Jan 12, 2018)

* Updating SDK version to v9.4.0 which is the latest stable to the date.
* Using caret range for SDK version.

## 1.4.1 (December 22, 2017)

* Updating SDK version to v9.3.7 which is the latest stable to the date.

## 1.4.0 (November 09, 2017)

* Internal code reordering. Moving to a routers structure.
* Moved `/machine` and `/version` to `/admin/machine` and `/admin/version`.
* Created `/admin/uptime` for consulting the current uptime of the server in a nice format.
* Created `/admin/ping` for ping calls to the server.
* Created `/admin/healthcheck` for checking the health of the evaluator.

## 1.3.1 (October 12, 2017)

* Updating SDK version to v9.3.6 which is the latest stable to the date.

## 1.3.0 (September 26, 2017)

* Rename to Split-Evaluator

## 1.2.2 (September 13, 2017)

Note: On this version we will start adding the patch version to the docker image tag,
so it will be 1.2.2 and of course, it will be in sync with callhome.

* Updating SDK version to v9.3.5 which is the latest stable to the date.

## 1.2.1 (September 8, 2017)

* Added `/machine/` endpoint.
  ```{
    ip: {ip_address},
    name: {machine_name}
  }```

## 1.2.0 (September 7, 2017)
Note: We are using 1.2.0 to start in order to match the version of the
docker image on Docker Hub, which is currently 1.2. From now on we will have those
versions in sync.

* Updating SDK version to v9.3.4 which is the latest stable to the date.
* Updated README.md.
* Added NEWS.txt, CHANGES.txt and LICENSE files.
* Removed docker-hub support.
* Making SPLITIO_EXT_API_KEY header optional. If it is not set, it won't be required.
* Adding some logs and errors for better debugging.
* Modifying `/get-treatment` response to return the `splitName` field.
  ```{
    "splitName": "my-experiment",
    "treatment": "on"
  }```
* Modifying `/get-treatments` endpoint to receive an array of keys and return only feature flags for
  the traffic types on which a key was provided.
  Example curl: curl 'http://localhost:4444/get-treatments?keys=\[\{"matchingKey":"my-first-key","trafficType":"account"\},\{"matchingKey":"my-second-key","bucketingKey":"my-bucketing-key","trafficType":"user"\}\]' -H 'Authorization: {SPLITIO_EXT_API_KEY}'
* Modifying `/get-treatments` response to be an array of the objects returned on get-treatment.
  ```[
    { "splitName": "my-experiment", "treatment":"on" },
    { "splitName": "my-second-experiment", "treatment":"off" },
    ...
    { "splitName": "my-last-experiment", "treatment":"visa" }
  ]```
* Modified `/version` response.
  ```{
    version: {callhome_version},
    sdk: {sdk_language},
    sdkVersion: {sdk_version_number}
  }```

## 1.1.0 (July 28, 2017)

* Adding Environment variables for debugging.
* Updating SDK version to v9.3.1.

## 1.0.0 (May 28, 2017)

* Initial public release.


routes/admin.js
/**
 * To be mounted at /admin
 */
const express = require('express');
const router = express.Router();

const os = require('os');
const ip = require('ip');

const utils = require('../utils');
const sdkModule = require('../sdk');
const sdk = sdkModule.factory;

/**
 * ping method, just that.
 */
router.get('/ping', (req, res) => {
  console.log('pong');
  res.status(200).send('pong');
});
/**
 * health check method. Will respond according to SDK status. Two options:
 *    200 - Everything is OK. Ready to evaluate.
 *    500 - Something is wrong or not ready yet. Not able to perform evaluations.
 */
router.get('/healthcheck', (req, res) => {
  console.log('Running health check.');
  let status = 500;
  let msg = 'Split evaluator engine is not evaluating traffic properly.';

  if (sdkModule.isReady()) {
    status = 200;
    msg = 'Split Evaluator working as expected.';
  }

  console.log('Health check status: ' + status + ' - ' + msg);
  res.status(status).send(msg);
});
/**
 * version method. Returns the version of the Split evaluator and the SDK type and version.
 */
router.get('/version', (req, res) => {
  console.log('Getting version.');
  const version = utils.getVersion();
  const parts = sdk.settings.version.split('-');
  const sdkLanguage = parts[0];
  const sdkVersion = parts.slice(1).join('-');

  res.send({
    version,
    sdk: sdkLanguage,
    sdkVersion
  });
});
/**
 * Returns some data from the machine running the evaluator, ip and hostname.
 */
router.get('/machine', (req, res) => {
  console.log('Getting machine information.');
  let address; let hostname;

  try {
    address = ip.address();
    hostname = os.hostname();
  } catch(e) {
    address = hostname = 'unavailable';
  }

  res.send({
    ip: address,
    name: hostname
  });
});
/**
 * Returns the uptime of the server.
 */
router.get('/uptime', (req, res) => {
  const uptime = utils.uptime();
  console.log('Getting uptime: ' + uptime);
  res.send('' + uptime);
});

module.exports = router;
